File: CNC modeling 2.0/src/main/java/com/sergey/pisarev/model/Expression.java
package com.sergey.pisarev.model;

import java.util.Stack;

public class Expression {
    //Метод возвращает true, если проверяемый символ - разделитель ("пробел" или "равно")
    private static boolean isDelimeter(char c) {
        return (" =".indexOf(c) != -1);
    }

    //Метод возвращает true, если проверяемый символ - оператор
    private static boolean isOperator(char с) {
        return ("+-/*^()".indexOf(с) != -1);
    }

    //Метод возвращает приоритет оператора
    private static byte getPriority(char s) {
        switch (s) {
            case '(':
                return 0;
            case ')':
                return 1;
            case '+':
                return 2;
            case '-':
                return 3;
            case '*':
            case '/':
                return 4;
            case '^':
                return 5;
            default:
                return 6;
        }
    }

    //"Входной" метод класса
    public float calculate(String input) {
        input = insertZero(input);
        String output = getExpression(input); //Преобразовываем выражение в постфиксную запись
        return counting(output); //Возвращаем результат
    }

    private String insertZero(String input) {
        StringBuilder sb = new StringBuilder(input);
        if (input.contains("-") && sb.charAt(0) == '-') {
            sb = sb.replace(0, 0, "0");
        } else if (input.contains("-") && sb.indexOf("-") != 0) {
            for (int i = 0; i < sb.length(); i++) {
                if (sb.charAt(i) == '-') {
                    if (sb.charAt(i - 1) == '(' || sb.charAt(i - 1) == '+' || sb.charAt(i - 1) == '-') {
                        sb = sb.replace(i, i, "0");
                    }
                }
            }
        } else if (input.contains("+") && sb.indexOf("+") != 0) {
            for (int i = 0; i < sb.length(); i++) {
                if (sb.charAt(i) == '+') {
                    if (sb.charAt(i - 1) == '(' || sb.charAt(i - 1) == '+' || sb.charAt(i - 1) == '-') {
                        sb = sb.replace(i, i, "0");
                    }
                }
            }
        }
        return sb.toString();
    }

    private static String getExpression(String input) {
        StringBuilder output = new StringBuilder(); //Строка для хранения выражения
        Stack<Character> operStack = new Stack<>(); //Стек для хранения операторо
        char[] inputArr = input.toCharArray();

        for (int i = 0; i < inputArr.length; i++) //Для каждого символа в входной строке
        {
            //Разделители пропускаем
            if (isDelimeter(inputArr[i]))
                continue; //Переходим к следующему символу

            //Если символ - цифра, то считываем все число
            if (Character.isDigit(inputArr[i])) //Если цифра
            {
                //Читаем до разделителя или оператора, что бы получить число
                while (!isDelimeter(inputArr[i]) && !isOperator(inputArr[i])) {
                    output.append(inputArr[i]); //Добавляем каждую цифру числа к нашей строке
                    i++; //Переходим к следующему символу

                    if (i == inputArr.length) break; //Если символ - последний, то выходим из цикла
                }

                output.append(" "); //Дописываем после числа пробел в строку с выражением
                i--; //Возвращаемся на один символ назад, к символу перед разделителем
            }

            //Если символ - оператор
            if (isOperator(inputArr[i])) //Если оператор
            {
                if (inputArr[i] == '(') //Если символ - открывающая скобка
                    operStack.push(inputArr[i]); //Записываем её в стек
                else if (inputArr[i] == ')') //Если символ - закрывающая скобка
                {
                    //Выписываем все операторы до открывающей скобки в строку
                    char s = operStack.pop();

                    while (s != '(') {
                        output.append(s).append(" ");
                        s = operStack.pop();
                    }
                } else //Если любой другой оператор
                {
                    if (operStack.size() > 0) //Если в стеке есть элементы
                        if (getPriority(inputArr[i]) <= getPriority(operStack.peek())) //И если приоритет нашего оператора меньше и
                            //ли равен приоритету оператора на вершине стека
                            output.append(operStack.pop()).append(" "); //То добавляем последний оператор из стека в строку с выражением

                    operStack.push(inputArr[i]); //Если стек пуст, или же приоритет оператора выше - добавляем операторов на вершину стека
                }
            }
        }
        //Когда прошли по всем символам, выкидываем из стека все оставшиеся там операторы в строку
        while (operStack.size() > 0)
            output.append(operStack.pop()).append(" ");
        return output.toString(); //Возвращаем выражение в постфиксной записи
    }

    private static float counting(String input) {
        char[] inputArr = input.toCharArray();
        float result = 0; //Результат
        Stack<Float> temp = new Stack<>(); //Dhtvtyysq стек для решения
        for (int i = 0; i < input.length(); i++) //Для каждого символа в строке
        {
            //Если символ - цифра, то читаем все число и записываем на вершину стека
            if (Character.isDigit(inputArr[i])) {
                StringBuilder a = new StringBuilder();
                while (!isDelimeter(inputArr[i]) && !isOperator(inputArr[i])) //Пока не разделитель
                {
                    a.append(inputArr[i]); //Добавляем
                    i++;
                    if (i == input.length()) break;
                }
                temp.push(Float.parseFloat(a.toString())); //Записываем в стек
                i--;
            } else if (isOperator(inputArr[i])) //Если символ - оператор
            {
                //Берем два последних значения из стека
                float a = temp.pop();
                float b = temp.pop();
                switch (inputArr[i]) //И производим над ними действие, согласно оператору
                {
                    case '+':
                        result = b + a;
                        break;
                    case '-':
                        result = b - a;
                        break;
                    case '*':
                        result = b * a;
                        break;
                    case '/':
                        result = b / a;
                        break;
                    case '^':
                        result = (float) Math.pow(b, a);
                        break;
                }
                temp.push(result); //Результат вычисления записываем обратно в стек
            }
        }
        return temp.peek(); //Забираем результат всех вычислений из стека и возвращаем его
    }
}

File: CNC modeling 2.0/src/main/java/com/sergey/pisarev/model/Expression.kt
package com.sergey.pisarev.model

import java.lang.StringBuilder
import java.util.*

object Expression {
    //Метод возвращает true, если проверяемый символ - разделитель ("пробел" или "равно")
    private fun isDelimeter(c: Char): Boolean {
        return " =".indexOf(c) != -1
    }

    //Метод возвращает true, если проверяемый символ - оператор
    private fun isOperator(с: Char): Boolean {
        return "+-/*^()".indexOf(с) != -1
    }

    //Метод возвращает приоритет оператора
    private fun getPriority(s: Char): Byte {
        return when (s) {
            '(' -> 0
            ')' -> 1
            '+' -> 2
            '-' -> 3
            '*', '/' -> 4
            '^' -> 5
            else -> 6
        }
    }

    //"Входной" метод класса
    fun calculate(input: String): Float {
        var input = input
        input = insertZero(input)
        val output = getExpression(input) //Преобразовываем выражение в постфиксную запись
        return counting(output) //Возвращаем результат
    }

    private fun insertZero(input: String): String {
        var sb = StringBuilder(input)
        if (input.contains("-") && sb[0] == '-') {
            sb = sb.replace(0, 0, "0")
        }
        if (input.contains("-") && sb.indexOf("-") != 0) {
            for (i in 0 until sb.length) {
                if (sb[i] == '-') {
                    if (sb[i - 1] == '(' || sb[i - 1] == '+' || sb[i - 1] == '-') {
                        sb = sb.replace(i, i, "0")
                    }
                }
            }
        }
        if (input.contains("+") && sb.indexOf("+") != 0) {
            for (i in 0 until sb.length) {
                if (sb[i] == '+') {
                    if (sb[i - 1] == '(' || sb[i - 1] == '+' || sb[i - 1] == '-') {
                        sb = sb.replace(i, i, "0")
                    }
                }
            }
        }
        return sb.toString()
    }

    private fun getExpression(input: String): String {
        val output = StringBuilder() //Строка для хранения выражения
        val operStack = Stack<Char>() //Стек для хранения операторо
        val inputArr = input.toCharArray()
        var i = 0
        while (i < inputArr.size) {

            //Разделители пропускаем
            if (isDelimeter(inputArr[i])) {
                i++
                continue  //Переходим к следующему символу
            }

            //Если символ - цифра, то считываем все число
            if (Character.isDigit(inputArr[i])) //Если цифра
            {
                //Читаем до разделителя или оператора, что бы получить число
                while (!isDelimeter(inputArr[i]) && !isOperator(inputArr[i])) {
                    output.append(inputArr[i]) //Добавляем каждую цифру числа к нашей строке
                    i++ //Переходим к следующему символу
                    if (i == inputArr.size) break //Если символ - последний, то выходим из цикла
                }
                output.append(" ") //Дописываем после числа пробел в строку с выражением
                i-- //Возвращаемся на один символ назад, к символу перед разделителем
            }

            //Если символ - оператор
            if (isOperator(inputArr[i])) //Если оператор
            {
                if (inputArr[i] == '(') //Если символ - открывающая скобка
                    operStack.push(inputArr[i]) //Записываем её в стек
                else if (inputArr[i] == ')') //Если символ - закрывающая скобка
                {
                    //Выписываем все операторы до открывающей скобки в строку
                    var s = operStack.pop()
                    while (s != '(') {
                        output.append(s).append(" ")
                        s = operStack.pop()
                    }
                } else  //Если любой другой оператор
                {
                    if (operStack.size > 0) //Если в стеке есть элементы
                        if (getPriority(inputArr[i]) <= getPriority(operStack.peek())) //И если приоритет нашего оператора меньше и
                        //ли равен приоритету оператора на вершине стека
                            output.append(operStack.pop()).append(" ") //То добавляем последний оператор из стека в строку с выражением
                    operStack.push(inputArr[i]) //Если стек пуст, или же приоритет оператора выше - добавляем операторов на вершину стека
                }
            }
            i++
        }
        //Когда прошли по всем символам, выкидываем из стека все оставшиеся там операторы в строку
        while (operStack.size > 0) output.append(operStack.pop()).append(" ")
        return output.toString() //Возвращаем выражение в постфиксной записи
    }

    private fun counting(input: String): Float {
        val inputArr = input.toCharArray()
        var result = 0f //Результат
        val temp = Stack<Float>() // стек для решения
        var i = 0
        while (i < input.length) {

            //Если символ - цифра, то читаем все число и записываем на вершину стека
            if (Character.isDigit(inputArr[i])) {
                val a = StringBuilder()
                while (!isDelimeter(inputArr[i]) && !isOperator(inputArr[i])) //Пока не разделитель
                {
                    a.append(inputArr[i]) //Добавляем
                    i++
                    if (i == input.length) break
                }
                temp.push(a.toString().toFloat()) //Записываем в стек
                i--
            } else if (isOperator(inputArr[i])) //Если символ - оператор
            {
                //Берем два последних значения из стека
                val a = temp.pop()
                val b = temp.pop()
                when (inputArr[i]) {
                    '+' -> result = b + a
                    '-' -> result = b - a
                    '*' -> result = b * a
                    '/' -> result = b / a
                    '^' -> result = Math.pow(b.toDouble(), a.toDouble()).toFloat()
                }
                temp.push(result) //Результат вычисления записываем обратно в стек
            }
            i++
        }
        return temp.peek() //Забираем результат всех вычислений из стека и возвращаем его
    }
}